teos/src/responder.rs

```diff
@@ -336,179 +250,92 @@ impl Responder {
         completed_trackers
     }
 
-    /// Gets a map of transactions that need to be rebroadcast. A [Transaction] is flagged to be rebroadcast
-    /// if its missed confirmation count has reached the threshold ([CONFIRMATIONS_BEFORE_RETRY]) or if they have been
-    /// reorged out of the chain. If the transaction has been reorged out, the commitment transaction is also returned.
+    /// Handles the reorged out trackers when we start connecting to the stronger chain.
     ///
-    /// Given the [Responder] only keeps around the minimal data to track transactions, the [TransactionTracker]s
-    /// are queried to the [DBM].
-    fn get_txs_to_rebroadcast(
-        &self,
-        height: u32,
-    ) -> HashMap<UUID, (Transaction, Option<Transaction>)> {
+    /// This is called in the first block connection after a bunch of block disconnections.
+    /// It marks previously confirmed trackers in the reorged chain as in-mempool till they appear in our txindex.
+    fn handle_reorged_txs(&self, height: u32) {
+        // NOTE: We are draining the reorged trackers vector, meaning that we won't try sending these disputes again.
+        let reorged_trackers: Vec<UUID> = self.reorged_trackers.lock().unwrap().drain(..).collect();
+        let mut carrier = self.carrier.lock().unwrap();
         let dbm = self.dbm.lock().unwrap();
-        let mut tx_to_rebroadcast = HashMap::new();
-        let mut tracker: TransactionTracker;
-
-        for (uuid, t) in self.trackers.lock().unwrap().iter() {
-            if let ConfirmationStatus::InMempoolSince(h) = t.status {
-                if (height - h) as u8 >= CONFIRMATIONS_BEFORE_RETRY {
-                    tracker = dbm.load_tracker(*uuid).unwrap();
-                    tx_to_rebroadcast.insert(*uuid, (tracker.penalty_tx, None));
-                }
-            } else if let ConfirmationStatus::ReorgedOut = t.status {
-                tracker = dbm.load_tracker(*uuid).unwrap();
-                tx_to_rebroadcast.insert(*uuid, (tracker.penalty_tx, Some(tracker.dispute_tx)));
-            }
-        }
 
-        tx_to_rebroadcast
-    }
+        // First, mark all these trackers as in-mempool because we haven't seen the block they are in yet (if any).
+        for uuid in reorged_trackers.iter() {
+            // We are marking them as last tried in `height - CONFIRMATIONS_BEFORE_RETRY` because we want them
+            // to be rebroadcasted ASAP (in this block connection).
+            dbm.update_tracker_status(
```

Comment:
I think this may be equivalent to what we were doing, but I also think we can do better. There are several things here that I'm not too convinced about:

- First, I don't think you have to go through all this now, given we really don't need to rebroadcast a penalty if the dispute is rejected.
- Furthermore, I don't really like the logic of setting this to be `InMempoolSince` with the proper height just so it triggers in this iteration of `rebroadcast_stale_txs`, that feels a bit hacky.

How about we flag reorged transactions in the database with confirmed = 0 and height = -1 (or any combination of values with the -1 special meaning) and we store those during `block_disconnected`, then, later when we are connecting blocks again, `load_trackers_with_confirmation_status` can load both reorged and stale transactions.

This kind of kills two birds with one stone. You won't need to keep track of anything reorged in memory, just load the trackers with the proper state for `handle_reorged_txs`. Also, instead of having to update all those states, you will leave them as they are, or delete them if the dispute gets rejected. Finally, you may update this when they are accepted to mempool.

Reply:
Notice this also works reworking the database to store the state as a number and the height as dependant to the state, instead of `confirmed` + `height`

Reply:
> Furthermore, I don't really like the logic of setting this to be InMempoolSince with the proper height just so it triggers in this iteration of rebroadcast_stale_txs, that feels a bit hacky.

Strongly agree!

> You won't need to keep track of anything reorged in memory, just load the trackers with the proper state for handle_reorged_txs.

This should be the way eventually.

> Notice this also works reworking the database to store the state as a number and the height as dependant to the state, instead of confirmed + height

That's why I'm hacking it out like that, using the current primitives we have. I want to keep the DB the same for now to avoid having to do manual migrations using the current DB engine we use (`rusqlite`).
When moving to [sqlx](https://github.com/launchbadge/sqlx) we will have the flexibility of altering the DB schema without losing all the running towers.

Reply:
I gave this another check, and even if we improve it later on by storing some data in the database, I still think this is mixing up things.

The old version of the code did the following regarding reorgs:

When blocks were being disconnected, it kept flagging everything that was added at the disconnected height as `ReorgedOut`, once blocks started to connect again, in the first newly connected block, it'll handle all the data marked as `ReorgedOut` and try to rebroadcast the `dispute` (if it was not found) and the `penalty` if the former was either found or accepted. Accepted data was flagged as `InMempoolSince(first_connected_height)` so data was not sent over and over again even if multiple blocks were connected one after the other. This may be repeated modulo `CONFIRMATIONS_BEFORE_RETRY` if we don't see the transactions being confirmed soon enough. 

However, now we splitting disputes and penalties, handling the former in `handle_reorged_txs` and the later in `rebroadcast_stale_txs`, where they are not really stale. Plus, we are hacking it by storing the penalties as rebroascastable in this iteration, which is pretty nasty. I think we should stick to the previous logic and handle the package of dispute + penalty in `handle_reorged_txs`, and leave `rebroadcast_stale_txs` for truly stale transactions.